### ISSUE TITLE
Question about qat_engine_init

### ISSUE THREAD

**Issue by konishi-av** [01org/QAT_Engine#1]
I have a question about a source of `e_qat.c`.
Can you please help me on this issue?

I want to develop `NGINX` which accelerated SSL processing using `QAT Engine` and `OpenSSL 1.1.0 pre5` by `EVENT_DRIVEN`.
It is described in a document as follows.:
 

> It must be sent if required after engine creation but before engine initialization. It should not be sent after engine initialization.

I think that `Engine initialization` is `Engine_init`.
Because `qat_engine_init` which is called by `Engine_init` finished all in the following logics, after I can not acquire `FD` using `GET_POLLIG_FD`.
```
/* Work around for fork case, finish engine in parent process */
if(initiator_pid == 0) {
    qat_engine_finish(e);
    qat_inited = 0;
    keep_polling = 1;
}
``` 
In a `qat_engine_init` source comment, I think that this considers `master/worker` process like  `NGINX`.

Because `qat_engine_init` is called from `CRYPTO_THREAD_run_once(&engine_lock_init, do_engine_lock_init);`, it is called only once.
In other words it is never called in worker because it is initialized in master.

How should I initialize it?

**Comment by stevelinsell** [comment]
Yes, I have also encountered this issue a couple of weeks ago. The development team are addressing it during this development cycle which has just started so if all goes to plan you should expect a fix to be posted to here within the next two weeks.
As you noted the issue is caused because NGINX initialises the engine in the master process but only uses the crypto functionality in the worker processes. The Intel Quickassist Technology Engine calls the qat_engine_finish() function in the master process so as not to waste a process (you can only use acceleration from a limited number of processes). The engine relies on the fact the get_next_inst() function within e_qat.c will be called on every crypto operation to trigger starting the engine again in any worker processes. Unfortunately the case where Event Driven is used was not considered as in order for the GET_POLLING_FD custom engine message to work and retrieve the FD there needs to be valid instances available which there are not.
A workaround for the issue (untested) would be to add a call to get_next_inst() at Line 1116 of e_qat.c https://github.com/01org/QAT_Engine/blob/master/e_qat.c#L1116
You would then need to ensure that the GET_POLLING_FD custom engine message was called in the worker processes and not in the master process. This would involve adding the call to send the custom engine message to ngx_worker_process_init() instead. It is fine to continue to send the ENABLE_EVENT_DRIVEN_MODE engine message in the master process before the engine is initialised via ENGINE_init().

It should also be noted that it is only necessary to use the GET_POLLING_FD message if you are using external polling (i.e. have sent the message ENABLE_POLLING) AND are using Event Driven Mode, (i.e. have set ENABLE_EVENT_DRIVEN_MODE). It is possible to just set ENABLE_EVENT_DRIVEN_MODE without setting ENABLE_POLLING. In this mode of operation an internal thread will be created within the engine that will handle the Event Driven Polling. It will still be necessary to make changes to the NGINX application in order to handle the Asynchronous mode at the libSSL level, i.e. handle the asynchronous status returned by SSL_get_error() and use SSL_get_changed_async_fds to get asynchronous FD's to listen for async events at the SSL connection level. The performance of using the internal thread for the polling will also be lower than using the external polling due to context switching between the main thread and the polling thread.

Let me know if you need further info.

**Comment by konishi-av** [comment]
Thank you for information. I will try it.

**Comment by konishi-av** [comment]
I execute `OpenSSL` with `ASYNC` mode, and `QAT` with `EVENT DRIVEN` mode.
And I call `poll` after waiting using `epoll_wait()`.

If `OpenSSL` functions (e.g. `SSL_read`, `SSL_write`) return `SSL_ERROR_WANT_ASYNC`, basically I call `poll` after waiting using `epoll_wait()`.
But even if after calling `poll`, I seem to have possibilities to freeze when I call `OpenSSL` functions.

This issue is solved if I call from `epoll_wait()` to `poll` in the other thread.
Is this specifications?

**Comment by stevelinsell** [comment]
Let me take a step back and explain the architecture for the different modes you can run in. Hopeful that will help you know whether what you are doing is the correct way.

I will start with the underlying Intel Quickassist Technology Userspace Driver which can work in different modes. It can work in **Polled Mode** (For each instance in the driver config file you have set `CyXIsPolled = 1`) or it can work in **Event Driven Mode** (For each instance in the driver config file you have set `CyXIsPolled = 2`).

In **Polled Mode** it is the users responsibility to periodically call `icp_sal_CyPollInstance` which will check for messages that have come back from the hardware and run the callbacks associated with those messages.

In **Event Driven Mode** a file descriptor is created for each instance. The file descriptors can be retrieved using the API call `icp_sal_CyGetFileDescriptor` on each instance. The file descriptor can then be placed in something like an epoll loop to listen for events. When an event is received it is then the users responsibility to call `icp_sal_CyPollInstance` to check for messages and run the callback. Note that these events on a file descriptor are at the instance level NOT an SSL connection level.

When you use the Intel Quickassist Technology OpenSSL Engine by default most of the above is taken care of for you within the engine. All you need to do is tell the engine if you are running in **Event Driven Mode** by sending the `ENABLE_EVENT_DRIVEN_MODE` custom engine message. The default is that you are running in **Polled Mode** in which case no custom engine message needs to be sent.
Within the engine the **Polled Mode** is taken care of by creating a thread for each instance that will periodically call `icp_sal_CyPollInstance` and run the callback associated with the message. If `ENABLE_EVENT_DRIVEN_MODE` has been set, then a single thread is created that will run an epoll loop containing each instances file descriptor, and will call `icp_sal_CyPollInstance` as appropriate.

When you send the Intel Quickassist Technology OpenSSL Engine the `ENABLE_POLLING` message you are telling the engine that you as the application want to take care of checking for messages rather than the engine. If this message is sent then no internal threads will be created within the engine. If you send that message then you are going to do one of two things in the application:

1. If you are running the Intel Quickassist Techonology Userspace Driver in **Polled Mode** then as an application you will need to periodically send the `POLL` custom engine message that will result in a call to `icp_sal_CyPollInstance`. You could send this engine message on a timer or in a separate thread but you must ensure that you don't end up in a deadlock situation where the application is waiting for a response at the SSL level but is not sending `POLL`.

2. If you are running the Intel Quickassist Technology Userspace Driver in **Event Driven Mode** then as an application you will need to use the `GET_POLLING_FD` custom engine message to retrieve the file descriptors for each instance. You would then add these to an epoll loop at the application level and if any events are triggered then you send the `POLL` custom engine message. You could do this within nginx by adding a dummy connection for each file descriptor and adding them to the main epoll loop, with the action of the event being to send the `POLL` message. Again it is important that the epoll is run regularly and does not allow a deadlock situation to arise.

This is all the special handling needed for using the Intel Quickassist Technology OpenSSL Engine but in order to make nginx run SSL connections in asynchronous fashion you will also need to change nginx to use the new asynchronous support of OpenSSL including using different file descriptors at the SSL connection level as notifiers of when you can make progress on a particular connection. This is documented in the OpenSSL .pod files.

**Comment by konishi-av** [comment]
Thank you for reply.
I will use your explanation as reference.

**Comment by agrandi** [comment]
Thank you @stevelinsell for answering the question!
@konishi-av feel free to re-open the issue if you need additional clarifications.

**Comment by konishi-av** [comment]
Hi all, excuse me, but

@stevelinsell wrote on 28 Jun: 

> Yes, I have also encountered this issue a couple of weeks ago. The development team are addressing it during this development cycle which has just started so if all goes to plan you should expect a fix to be posted to here within the next two weeks.

When is the issue mentioned above fixed?
Please let me know the schedule.

Kind regards,

@konishi-av
